dora/core/server/worker/src/main/java/alluxio/worker/s3/S3NettyHandler.java

```diff
@@ -0,0 +1,705 @@
+    String pathStr = path;
+    String bucket = null;
+    String object = null;
+    S3NettyHandler handler = null;
+    try {
+      if (bucketMatcher.matches()) {
+        pathStr = java.net.URI.create(path.substring(1)).getPath();
+        bucket = URLDecoder.decode(pathStr, "UTF-8");
+      } else if (objectMatcher.matches()) {
+        pathStr = java.net.URI.create(path.substring(1)).getPath();
+        bucket = URLDecoder.decode(
+            pathStr.substring(0, pathStr.indexOf(AlluxioURI.SEPARATOR)), "UTF-8");
+        object = URLDecoder.decode(
+            pathStr.substring(pathStr.indexOf(AlluxioURI.SEPARATOR) + 1), "UTF-8");
+      }
+      handler = new S3NettyHandler(bucket, object, request, context, fileSystem, doraWorker,
+          asyncAuditLogWriter);
+      handler.setStopwatch(stopwatch);
+      handler.init();
+      S3NettyBaseTask task = null;
+      if (object != null && !object.isEmpty()) {
+        task = S3NettyObjectTask.Factory.create(handler);
+      } else {
+        task = S3NettyBucketTask.Factory.create(handler);
+      }
+      handler.setS3Task(task);
+      return handler;
+    } catch (Exception ex) {
+      LOG.error("Exception during create s3handler:{}", ThreadUtils.formatStackTrace(ex));
+      throw ex;
+    }
+  }
+
+  /**
+   * Initialize the S3Handler object in preparation for handling the request.
+   * @throws Exception
+   */
+  public void init() throws Exception {
+    // Do Authentication of the request.
+    doAuthentication();
+    // Extract x-amz- headers.
+    extractAMZHeaders();
+    // Reject unsupported subresources.
+    rejectUnsupportedResources();
+    // Init utils
+
+    // TODO(wyy) init directories
+    // Initiate the S3 API MPU metadata directories
+  }
+
+  /**
+   * Utility function to help extract x-amz- headers from request.
+   */
+  public void extractAMZHeaders() {
+    Iterator<Map.Entry<String, String>> headerIt = mRequest.headers().iteratorAsString();
+    while (headerIt.hasNext()) {
+      Map.Entry<String, String> header = headerIt.next();
+      mAmzHeaderMap.putIfAbsent(header.getKey(), header.getValue());
+    }
+  }
+
+  /**
+   * Reject unsupported request from the given subresources from request.
+   * @throws S3Exception
+   */
+  public void rejectUnsupportedResources() throws S3Exception {
+    String uri = mRequest.uri();
+    QueryStringDecoder decoder = new QueryStringDecoder(uri);
+    Map<String, List<String>> parameters = decoder.parameters();
+    for (String parameter : parameters.keySet()) {
+      if (mUnsupportedSubResourcesSet.contains(parameter)) {
+        throw new S3Exception(S3Constants.EMPTY, S3ErrorCode.NOT_IMPLEMENTED);
+      }
+    }
+  }
+
+  /**
+   * Do S3 request authentication.
+   * @throws Exception
+   */
+  public void doAuthentication() throws Exception {
+    try {
+      mUser = NettyRestUtils.getUser(mRequest);
+    } catch (Exception e) {
+      LOG.warn("exception happened in Authentication.");
+      throw e;
```

Comment:
LOG and rethrow is an anti-pattern

Reply:
Do not catch and log exception here. Print this exception log in where to actually handle the exception.